### DISCUSSION TITLE
Getting extruder position over time

### DISCUSSION THREAD

**Discussion by adamwolf** [vintagepc/MK404#366] [Q&A]
Hi!

This is awesome. Could I use this to get a list of coordinates of the extruder tip?  Would I have to watch the step/dir pins per axis, or is this already calculated?

**Comment by vintagepc** [comment]
Hi there!

It's not currently exposed in a direct manner, but it should be relatively straightforward to do, with a few points of note: 

The `Trace` system  (`-t` on the command line) is able to capture IRQ values and dump them to a .VCD file. Many of the stepper functions (step/dir) as you note are already captured; the gotcha is this runs at a fixed sampling frequency rather than capturing every change - so if per-step accuracy is what you are looking for it may not be quite what you are looking for. (this frequency can be changed though)

If that sounds reasonable it's just a 1-line change to instruct it to also capture the position, in "steps" which you can then convert based on steps/mm and microstep resolutions. (I wouldn't necessarly recommend capturing the raw floating-point positions, as there is some hackery to mash that into a uint32 and back again for use with the integer-based IRQ system.)

Alternately, the OpenGL helper class (MK3SGL.cpp) has all of these values come to a single location since it needs to know them to draw the 3D models - so this is potentially another place where some code could be added to dump data to a file. 

Yet another possible approach is the "PLY export" format. The original intent of this was to be able to capture/dump simulated prints from the 3D view for future reference - however, I would think it could easily be repurposed - if there is constant positive extrusion on the E-axis (assuming you only care about XYZ), it will continue logging positions into the polygon data to be dumped. 

Can I maybe have a little more detail about what you're trying to do/the reason you'd like this data? Perhaps then I can suggest which is more suitable. 

Also, what is your programming skill level? (Just so I can craft my response more accordingly). 

Cheers, 
~Vintage

**Comment by adamwolf** [comment]
Thanks for the response!

I'm looking to run gcode and look for printhead/print collisions.

I'm a decent programmer but haven't slung much C++ in a while, so depending
on "how C++" this project is, I may have some more questions :)

**Comment by vintagepc** [comment]
Ah, nice. That's probably on a slow enough scale that the trace system is sufficient. 

This: 
https://github.com/vintagepc/MK404/blob/edb79ceb9917ab5def9a338ffbcc8bf67f84fc89/parts/components/TMC2130.cpp#L313 is the area you'll want to be in, I think it should be as simple as adding

`TH.AddTrace(this, STEP_POS_OUT, {TC::Stepper, TC::Misc});`

to the end of the list. 

You'll know if it worked if you see it show up in `./MK404 -t ?` and if it does you should then be able to call it out by name in the `--trace` argument.  If you need to increase the trace rate there is the `--tracerate` option, value is in microseconds of sim (not wall) time. 

Tip: Traces don't start automatically to reduce spam. Logging is started/stopped with either the `+`/`-` keys, or using the middle-click menu  `TelHost->[Start/Stop]Trace`

Hope that helps - and don't hesitate to ask if you need further help!
